metal/resource_metal_device.go

```diff
@@ -524,6 +546,8 @@ func resourceMetalDeviceCreate(d *schema.ResourceData, meta interface{}) error {
 		return retErr
 	}
 
+	// Pick the hostname validated by metal
+	d.Set("hostname", newDevice.Hostname)
```

Comment:
I don't setting the hostname here is needed. SSH provisioners will not execute until after waitForActiveDevice and resourceMetalDeviceRead (which picks up the hostname) below.

Reply:
In the first part of this function, the desired hostname is computed and then passed to the Metal API to create the new device.

Then here we sync back the hostname attribute with what we got back from the Metal API.

That way the Metal API can decide to name the hostname otherwise, and the resource hostname attribute is computed based on that.